core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,126 @@
+    this.lockCleaner =
+        new ScheduledThreadPoolExecutor(
+            1,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("tree-lock-cleaner-%d")
+                .build());
+
+    lockCleaner.scheduleAtFixedRate(
+        () -> {
+          long nodeCount = totalNodeCount.get();
+          LOG.info("Total tree lock node count: {}", nodeCount);
+          if (nodeCount > MAX_TREE_NODE_IN_MEMORY) {
+            treeLockRootNode
+                .getAllChildren()
+                .forEach(child -> evictStaleNodes(child, treeLockRootNode));
+          }
+        },
+        120,
+        5,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Evict the stale nodes from the tree lock node.
+   *
+   * @param treeNode The tree lock node to evict.
+   * @param parent The parent of the tree lock node.
+   */
+  @VisibleForTesting
+  void evictStaleNodes(TreeLockNode treeNode, TreeLockNode parent) {
+    // We will not evict the root node if the total node count is less than the
+    // MIN_TREE_NODE_IN_MEMORY.
+    // Do not need to consider thread-safe issues.
+    if (totalNodeCount.get() < MIN_TREE_NODE_IN_MEMORY) {
+      return;
+    }
+
+    // Handle from leaf nodes first.
+    treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+
+    // Handle self node.
+    if (treeNode.getReferenceCount() == 0) {
+      synchronized (parent) {
+        // Once goes here, the parent node has been locked, so the reference could not be changed.
+        if (treeNode.getReferenceCount() == 0) {
+          parent.removeChild(treeNode.getIdent());
+          totalNodeCount.decrementAndGet();
+          LOG.info("Evict stale tree lock node '{}' and all its children", treeNode.getIdent());
+        } else {
+          treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+        }
+      }
+    }
+  }
+
+  /**
+   * Create a tree lock with the given identifier.
+   *
+   * @param identifier The identifier of the tree lock.
+   * @return The created tree lock.
+   */
+  public TreeLock createTreeLock(NameIdentifier identifier) {
+    TreeLockNode lockNode = treeLockRootNode;
+    treeLockRootNode.addReference();
+
+    List<TreeLockNode> treeLockNodes = Lists.newArrayList(lockNode);
+    if (identifier.equals(NameIdentifier.ofRoot())) {
+      // The lock tree root node
+      return new TreeLock(treeLockNodes);
+    }
+
+    String[] levels = identifier.namespace().levels();
+    levels = ArrayUtils.add(levels, identifier.name());
+
+    try {
+      for (int i = 0; i < levels.length; i++) {
+        NameIdentifier ident = NameIdentifier.of(ArrayUtils.subarray(levels, 0, i + 1));
+        lockNode = lockNode.getOrCreateChild(ident);
+        treeLockNodes.add(lockNode);
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to create tree lock {}", identifier, e);
+      throw e;
+    }
```

Comment:
What will be the exception here? If there's an exception, should you decrease the ref count you previously added? Otherwise, it will be wrong.

Reply:
Got